File: src/main/java/com/hey/lec04/Lec04Main.java
package com.hey.lec04;

public class Lec04Main {

  public static void main(String[] args) {
    JavaMoney money1 = new JavaMoney(1_000L);
    JavaMoney money2 = new JavaMoney(2_000L);
    if (money1.compareTo(money2) > 0) {
      System.out.println("Money1이 Money2보다 금액이 큽니다");
    }
  }

  public static void identityEquality() {
    JavaMoney money1 = new JavaMoney(1_000L);
    JavaMoney money2 = money1;
    JavaMoney money3 = new JavaMoney(1_000L);

    System.out.println(money1 == money2); // true
    System.out.println(money1 == money3); // false
    System.out.println(money1.equals(money3)); // true
  }

  public static void plusObject() {
    JavaMoney money1 = new JavaMoney(1_000L);
    JavaMoney money2 = new JavaMoney(1_000L);
    System.out.println(money1.plus(money2));
  }

}

File: src/main/kotlin/com/hey/lec04/Lec04Main.kt
package com.hey.lec04

fun main() {
    val money1 = JavaMoney(2_000L)
    val money2 = JavaMoney(1_000L)

    if (money1 > money2) { // 비교연산자를 사용하면 compareTo를 사용한다
        println("Money1이 Money2보다 금액이 큽니다")
    }

    // lazy 연산 수행
    if (lazyFun1() || lazyFun2()) { // lazyFun1이 true이면 lazyFun2를 실행하지 않음
        println("true")
    }

    overLoading()
}

// 동등성과 동일성
fun identityEquality() {
    val money1 = JavaMoney(1_000L)
    val money2 = money1
    val money3 = JavaMoney(1_000L)

    println(money1 == money3) // true 값만 비교
    println(money1 === money3) // false 주소까지 비교
}

fun lazyFun1(): Boolean {
    println("lazyFun1")
    return true
}

fun lazyFun2(): Boolean {
    println("lazyFun2")
    return false
}

fun overLoading() {
    val money1 = JavaMoney(1_000L)
    val money2 = JavaMoney(1_000L)
    println(money1 + money2)
}
